sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalysisContext.java

```diff
@@ -22,27 +22,71 @@
 package io.crate.analyze.expressions;
 
 import io.crate.analyze.symbol.Function;
+import io.crate.analyze.symbol.Literal;
 import io.crate.analyze.symbol.Symbol;
+import io.crate.analyze.symbol.Symbols;
+import io.crate.metadata.FunctionImplementation;
 import io.crate.metadata.FunctionInfo;
+import io.crate.metadata.Functions;
+import io.crate.metadata.TransactionContext;
 import io.crate.sql.tree.ArrayComparisonExpression;
 import io.crate.sql.tree.Expression;
 
 import java.util.IdentityHashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 
+/**
+ * Base class for providing contextual information during expression analysis in
+ * {@link ExpressionAnalyzer}.
+ *
+ * 1) Function creation and folding constants (through symbol normalization),
+ *    e.g. turns '1 + 2' into 'Literal.of(3)'.
+ *
+ * 2) Detect sub queries in array expressions like '= ANY(Select id from t)'.
+ *
+ * 3) Indicates whether aggregates have been found during the analysis.
+ */
 public class ExpressionAnalysisContext {
 
     private final Map<Expression, Object> arrayExpressionsChildren = new IdentityHashMap<>();
 
-    public boolean hasAggregates = false;
+    private final Functions functions;
+    private final TransactionContext transactionContext;
+
+    private boolean hasAggregates;
 
-    Function allocateFunction(FunctionInfo functionInfo, List<Symbol> arguments) {
+    public ExpressionAnalysisContext(Functions functions, TransactionContext transactionContext) {
+        this.functions = Objects.requireNonNull(functions,
+            "Built-in functions may not be null");
+        this.transactionContext = Objects.requireNonNull(transactionContext,
+            "TransactionContext may not be null");
+    }
+
+    /**
+     * Tries to call normalize on the new function's {@link FunctionImplementation} iff
+     * only Literals are supplied as arguments. This folds any constant expressions like '1 + 1' => '2'.
+     * @param functionInfo The meta information about the new function.
+     * @param arguments The arguments to provide to the {@link Function}.
+     * @return The supplied {@link Function} or a {@link Literal} in case of constant folding.
+     */
+    Symbol allocateFunction(FunctionInfo functionInfo, List<Symbol> arguments) {
+        if (functionInfo.type() == FunctionInfo.Type.AGGREGATE) {
+            hasAggregates = true;
+        }
         Function newFunction = new Function(functionInfo, arguments);
-        hasAggregates = hasAggregates || functionInfo.type() == FunctionInfo.Type.AGGREGATE;
+        if (functionInfo.isDeterministic() && Symbols.allLiterals(newFunction)) {
+            FunctionImplementation funcImpl = functions.getQualified(newFunction.info().ident());
```

Comment:
Wouldn't it be cleaner to keep this kind of functionality in the
ExpressionAnalyzer? 

So far the ExpressionAnalysisContext was only responsible for "remembering" or
"holding onto" information. The ExpressionAnalyzer did the actual "work".

Reply:
a) The `ExpressionAnalysisContext` is directly bound to `ExpressionAnalyzer` since it is the only implementation.
b) The `allocateFunction` method already does some operations which could be part of `ExpressionAnalyzer`.

I suppose it would be cleaner to factor out everything in the context which is not state-related.